### PR TITLE
refactor(menu): rewrite component

### PR DESCRIPTION
**Pull request by sjohnson-sage**
Fixes: #3286
Fixes: #3343 

### Proposed behaviour
A rewrite of the `Menu` and `MenuItem` components. The submenu functionality of `MenuItem` has been split out into another component, `Submenu` without changing the interface for consumers and therefore avoiding a breaking change.

Rather than using props to pass things down from `Menu` to `MenuItem` it has now been refactored to use context instead, which means that `MenuItem` no longer needs to be a direct child of `Menu` in order for keyboard navigation etc to work. 

I have also split out as much of the keyboard navigation as I could into a separate file in an effort to simply the logic in these components.

Both `Menu` and `MenuItem` now extend the `Box` component, so all of the `Box` props can be added for alignment of children etc. An example of this is shown at the new `design-system-menu--full-navbar-alignment-example` story.

A "submenu direction left` story has also been added

### Current behaviour
See issue #3286 for current behaviour using props rather than context.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
A nice example to test keyboard navigation etc can be found here at the PR sandbox in the codesandbox bot comment below 
https://codesandbox.io/s/quizzical-rain-xns67?file=/src/index.js

### CODE REVIEW

File: src/components/menu/menu.component.js
Comment (edleeks87) [inline src/components/menu/menu.component.js:14]: This could be moved inside the scope of `onClickOutside` as it's not used anywhere else

**Review by edleeks87 — COMMENTED** [review]
You'll need to update or add some TypeScript definitions as a new component has been added and some prop types have been tightened

File: src/components/menu/submenu-block/submenu-block.component.js
Comment (edleeks87) [inline src/components/menu/submenu-block/submenu-block.component.js:16]: As mentioned in private conversation, this needs updating

File: cypress/locators/menu/index.js
Comment (sjohnson-sage) [reply cypress/locators/menu/index.js:12]: Ah yes, I've got lazy now prettier has been added in so I don't even bother typing one anymore, forgot it doesn't run over these files

**Comment by sjohnson-sage** [comment]
> You'll need to update or add some TypeScript definitions as a new component has been added and some prop types have been tightened

The new component is just an internal one so I don't think we need to provide a ts interface for it, I could be wrong about that though

File: src/components/menu/menu.stories.mdx
Comment (sjohnson-sage) [reply src/components/menu/menu.stories.mdx:355]: I've forgotten to add a `href` or `onClick` to those `MenuItem`s - I'll fix

File: src/components/menu/__internal__/submenu/submenu.component.js
Comment (sjohnson-sage) [reply src/components/menu/__internal__/submenu/submenu.component.js:88]: It could be but it's cleaner IMO to have this, easier to someone to look through and immediately see what's going on

File: src/components/menu/menu.component.js
Comment (jamime) [inline src/components/menu/menu.component.js:36]: You can test this, see ActionPopover spec for example.
```suggestion
```

File: src/components/menu/menu.component.js
Comment (jamime) [inline src/components/menu/menu.component.js:69]: Can this be simplified? Seems like we're passing it down, using `openSubmenu` but then passing it back as an argument to this function.
```suggestion
              handleKeyDown: ev =>
```

File: src/components/menu/submenu-block/submenu-block.component.js
Comment (jamime) [reply src/components/menu/submenu-block/submenu-block.component.js:16]: Do we need this warning at all?

**Review by jamime — CHANGES_REQUESTED** [review]
Good work sam.

One thing I noticed was there are two props `variant` and `menuType`, given we already have a codemod that can change a prop. I think it would be a good time to consolidate these by removing `menuType`.

File: src/components/menu/submenu-block/submenu-block.component.js
Comment (edleeks87) [reply src/components/menu/submenu-block/submenu-block.component.js:16]: The component may have been used even though it is stripped away inside `MenuItem`

**Comment by edleeks87** [comment]
> > You'll need to update or add some TypeScript definitions as a new component has been added and some prop types have been tightened
> 
> The new component is just an internal one so I don't think we need to provide a ts interface for it, I could be wrong about that though. Also I think the only other props I've changed are removing private ones, which weren't in the ts files to begin with

Ah yeah good point about the `__internal__`, I'd also missed you were updating `children` in `Menu` to be required as they are defined in the ts file

File: cypress/locators/menu/index.js
Comment (sjohnson-sage) [reply cypress/locators/menu/index.js:12]: Actually no I didn't, that's not the end of the line 😉

File: src/components/menu/__internal__/submenu/submenu.component.js
Comment (edleeks87) [reply src/components/menu/__internal__/submenu/submenu.component.js:88]: I don't think an `if/else` is so complex that someone couldn't work it out, but as there are a lot of nested ifs I can see your point about legibility so happy to defer to you here

**Comment by sjohnson-sage** [comment]
> Good work @sjohnson-sage.
> 
> One thing I noticed was there are two props `variant` and `menuType`, given we already have a codemod that can change a prop. I think it would be a good time to consolidate these by removing `menuType`.

Thanks @jamime! 

Do we want to do that though? `menuType` is set only on `Menu` to set the overall colour scheme, then `MenuItem` can override that, so we could have some menu items just the light theme, and some alternate as in the dropdown on this story: https://carbon.sage.com/?path=/story/design-system-menu--default-theme-alternate-colour 

I might be missing something but I think it makes sense as it is

File: src/components/menu/menu.component.js
Comment (sjohnson-sage) [reply src/components/menu/menu.component.js:69]: I can't see a way - this is specifically for when you have a submenu item focus and you press the left or right key onto another submenu. Somehow one submenu needs to communicate to open the second submenu through Menu, and this was the cleanest way I could fine. 
Happy to change if there's a better way I can't see though!

**Comment by jamime** [comment]
> > Good work @sjohnson-sage.
> > One thing I noticed was there are two props `variant` and `menuType`, given we already have a codemod that can change a prop. I think it would be a good time to consolidate these by removing `menuType`.
> 
> Thanks @jamime!
> 
> Do we want to do that though? `menuType` is set only on `Menu` to set the overall colour scheme, then `MenuItem` can override that with the `variant` prop, so we could have some menu items just the light theme, and some alternate as in the dropdown on this story: https://carbon.sage.com/?path=/story/design-system-menu--default-theme-alternate-colour
> 
> I might be missing something but I think it makes sense as it is

We've discussed and we're going to push this down the road to avoid the breaking change. We'll revisit this, likely when we do the DLS alignment exercise.

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Comment by sianford** [comment]
Tested the refactoring work and regression tested all of Menu in Safari, Chrome, Firefox and Edge Chromium in the Chromatic published Storybook. Testing passed ✅ Awaiting completion of the associated MDX example task before merging.

**Review by jamime — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by Alec-Walker-Sage — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by Alec-Walker-Sage — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Pull request merged**
